Better query builder page

heralden opened the pull request:
Depends on https://github.com/intermine/imcljs/pull/42.

A makeover with new features of the query builder page.
- Fixes #471
- Fixes #470 
- Fixes #468
- Recent queries
- Data browser
- Import from XML
- And lots of smaller improvements

Following needs to be done before ready to be merged.
- [x] ~~Support creating templates~~ Will do as separate PR
- [x] Test on older InterMine instances

heralden committed:
Improve look and responsiveness of querybuilder page

heralden committed:
Add Recent queries and skeleton components for in-progress features

heralden committed:
Use intent and query hash to implement recent queries

heralden committed:
Make model browser prettier and add data browser

heralden committed:
Clear im-query when setting root class

Fixes XML preview showing old query, and recent query still showing as
active, when changing root class.

heralden committed:
QB: Copy preferred root classes to top of dropdown

heralden committed:
QB preview: Scroll bars and increase font-size

heralden committed:
QB: Truncate paths in recent queries viewer to last two items

heralden committed:
QB: Implement saving of queries and Saved Queries tab

heralden committed:
QB: Sort saved queries by title

heralden committed:
QB: Allow undo of deleting saved query via message

heralden committed:
QB: Support importing of PathQuery XML

heralden committed:
QB: Handle importing invalid PathQuery XML

heralden committed:
QB: Cleanup sortable-list component

heralden committed:
QB: Support setting sortOrder of columns

heralden committed:
QB: Support outer join in query editor

heralden committed:
Fix utils/read-xml-query failing to read long paths

heralden committed:
QB: Referenced class and parent class, update outer join

Different title text and icons for outer join button.

heralden committed:
QB: List outer joins below query tree editor

heralden committed:
QB: Show constraint codes and improve delete icon placement

heralden committed:
QB: Add some top margin to joins list

heralden committed:
QB: Do not allow sorting of outerjoined sections

heralden committed:
QB: Fix long paths making query editor and columns ugly

heralden commented:
@yochannah Can you review? I just realized there's a feature here I need to use in a different PR. 

I will add template creation support in a different PR. I will also test this on older mines myself.

Please let me handle any merge conflicts if they're nontrivial.

Commit:
Merge branch 'dev' into better-query-builder-page

yochannah commented on less/components/querybuilder.less:
```diff
@@ -26,24 +26,183 @@ div.column-container {
+                cursor: pointer;
+
+                &.active-query {
+                    background-color: #f0fae5 !important;
```
!important often opens cans of worms for future maintenance - any chance of eliminating this?

heralden replied:
I remember struggling with this, failing to create a selector with a higher specificity than the others. I think the usage of !important here is correct since we want the active state to override all the others.

heralden replied:
^ This is wrong. I just needed to override the :hover pseudoclass [=

yochannah commented on less/components/querybuilder.less:
```diff
@@ -278,11 +443,55 @@ div.column-container {
+    }
+
+    .constraint-selector {
+        flex-direction: row !important;
```
as above - with a more specific selector we _should_ be able to manage without important, I think.

yochannah commented on less/components/querybuilder.less:
```diff
@@ -543,6 +757,55 @@ div.column-container {
+        .sort-button {
+            background-color: white;
+            font-size: 16px;
+            outline: solid 1px #dadada;
```
we have `#dadada`, `#fbfbfb` and `#d5d5d5` in a few places - if there aren't important visual distinctions between them, can we refactor to a variable?

heralden replied:
I changed `#dadada` to use `#d5d5d5` and refactored that to a variable. `#fbfbfb` is completely different and only used in one place, so I left that as it is.

yochannah replied:
👍

yochannah commented on less/components/querybuilder.less:
```diff
@@ -543,6 +757,55 @@ div.column-container {
+            flex-wrap: wrap;
+
+            .part {
+                padding: 0 2px;
```
spacer for padding and margin where possible would be nice :)

heralden replied:
Can we discuss the benefits of using `@spacer` when you're back? There are 40 lines that use margin/padding with px, so it will take me a long time to convert them without breaking alignment (since spacer is 1em, which changes depending on context, meaning I have to use the devtools and my very weary eyes to judge whether it's equivalent, for every single case).

yochannah commented on src/cljs/bluegenes/pages/querybuilder/events.cljs:
```diff
@@ -398,11 +444,16 @@
               :im-query nil
               :menu {})))
 
+(defn enhance-constraint-logic [logic]
```
could we add a docstring explaining why this is useful?

yochannah reviewed (changes requested):
Hey @uosl - this is great! Some comments inline, and also... a really long list. Sorry 😭 - I genuinely love the improvements you've already made, but few battle plans survive meeting the yo-enemy. 

-----

on my screen, lein prod produces this overlapping text in the model browser: 
<img width="496" alt="image" src="https://user-images.githubusercontent.com/9271438/74857034-04fbad00-533b-11ea-92ed-3846960b6f9c.png"> - same on Chrome and FF.

The sort columns were very low contrast - I can't see the icon very well unless it is clicked on (I have some mild visual impairment and glasses don't fix it perfectly. Others will probably have this complaint too).  
<img width="315" alt="image" src="https://user-images.githubusercontent.com/9271438/74857529-b39fed80-533b-11ea-8e1e-c98c72a601ab.png">

In manage columns/query editor, make sure to wrap the buttons and saved query name to the container width:

<img width="797" alt="image" src="https://user-images.githubusercontent.com/9271438/74857708-feba0080-533b-11ea-83af-fb38f0057ddd.png">
and
<img width="880" alt="image" src="https://user-images.githubusercontent.com/9271438/74857759-0da0b300-533c-11ea-8b0b-be15a2f69aec.png">

- If you save a query, I suggest switching to the saved query tab at the bottom. I nearly reported that it failed silently, but then realised I could switch tabs to see it. Alternatively (or maybe as well) a "toast" style notification to let us know it was saved would be good. 
- saved queries when I'm not logged in are transient and will die with the session, right? It would be nice to warn people. Either that or I'm saving them when I'm not logged in and I shouldn't be - not sure which 😆 
- when I click "load" on a saved query, I guess it loads in the query editor window? We should probably auto-focus that tab when we click load. When I have manage columns, or when I'm scrolled down, it's hard to tell if it did anything. Screenshot where I clicked load but see no change: 
<img width="1275" alt="image" src="https://user-images.githubusercontent.com/9271438/74858159-c2d36b00-533c-11ea-8cd4-9c73916b4a63.png"> 
- same for the "import xml" load query behaviour - the xml disappears but I don't see any change on the screen when I click load query and have a different tab open or am scrolled down too far. 
- should create template appear when I am not logged in? 
![image](https://user-images.githubusercontent.com/9271438/74858864-eba83000-533d-11ea-98b6-4cae0d7bf2b7.png) - oh wait, this template feature is still to be done, isn't it? never mind



----

Longer term wishlist - should I create a ticket? These shouldn't block the merge.
- on the xml tab I wished for a "copy" button.
- failing that, I clicked on the xml and pressed CTRL+A. This selected the whole page - maybe a textarea would be better?

heralden committed:
Merge remote-tracking branch 'upstream/dev' into better-query-builder-page

heralden commented:
![](https://user-images.githubusercontent.com/9271438/74857034-04fbad00-533b-11ea-92ed-3846960b6f9c.png)

I'll need help with this one. Already spent many hours on it before the commit, and I thought adding `height: 100%;` to all `li` elements fixed it. Well it did actually fix it, I have no idea why it reappeared now. Maybe it relied on some cached CSS that's now gone.

heralden committed:
QB: Greatly increase contrast of sort-button and refactor CSS variable

heralden committed:
QB: Wrap query editor buttons when narrow container

heralden commented:
> - when I click "load" on a saved query, I guess it loads in the query editor window? We should probably auto-focus that tab when we click load. When I have manage columns, or when I'm scrolled down, it's hard to tell if it did anything.
> - same for the "import xml" load query behaviour - the xml disappears but I don't see any change on the screen when I click load query and have a different tab open or am scrolled down too far. 

Loading a query can change the contents of the model browser, query editor and manage columns, depending on the differences between your active query and the loaded query. I don't think switching to the query editor tab would be the solution, since it's possible that the only thing different with the query you loaded is the sorting or header order (in which case it would only be visible in *Manage Columns*).

I don't know how to make it obvious that you've loaded a query, but I'm open to new ideas. A toast might work, but what text would we put in it? It should say something specific to the query loaded, but recent queries and most query xml don't have a title.

Edit: Recent and saved queries now turn green when they're loaded. Loading an XML query will now show a success message besides the button. Perhaps this is enough as a visual cue?

heralden committed:
QB: Message when saving query and warning if you're not logged in

heralden committed:
Fix subsequent messages not timing out by creating :message fx

When there were multiple messages active, the first one would time out
correctly and disappear, while the remaining would linger indefinitely.
Likely caused by lack of React keys combined with using didMount.

heralden committed:
Refactor duplicate CSS definitions of p.success and p.failure

heralden committed:
QB: Make XML preview more readable with button to copy

heralden committed:
QB: Remove uses of !important in CSS

heralden committed:
QB: Fix saved queries not highlighted when active

heralden committed:
QB: Docstring for enhance-constraint-logic

heralden commented:
I have implemented the remaining things you mentioned, that I haven't commented on.

Don't worry about being picky; many of these things are useful features and necessary fixes!

heralden committed:
QB: Success message when importing XML

yochannah commented on src/cljs/bluegenes/pages/querybuilder/events.cljs:
```diff
@@ -355,6 +370,37 @@
                      :order new-order)
       :dispatch [:qb/enhance-query-build-im-query true]})))
 
+(defn subpath-of? [subpath path]
+  (let [subpaths (->> (split path #"\.")
+                      (iterate drop-last)
+                      (take-while not-empty)
+                      (set))]
+    (contains? subpaths (split subpath #"\."))))
```
not a blocker for merge, but should this be in imcljs or better here? 🌷

heralden replied:
I thought long about the same... Can't remember what tipped me over to leave it out of imcljs.

yochannah committed:
lein format files

yochannah committed:
make overlaps in querybuilder model browser less overlappous.

yochannah reviewed (approved):
Okay - please merge once you've looked at my push and are happy to merge 👍 fabulous job!!

heralden committed:
QB: Rename colspan to colSpan to fix warning

heralden committed:
QB: Hide Create Template tab as it hasn't been implemented

heralden committed:
QB: Increase browser node height to avoid truncating tall letters

heralden committed:
QB: Fix incorrect root for saved queries on some InterMine

im-path/class was not the correct function for this, as it returns the
last class instead of the root.
